boot/init/tasks/switch_root.rb

```diff
@@ -145,10 +150,83 @@ def selected_generation()
     @selected_generation
   end
 
+  # If the system build exports a DTB file name to load, return the appropriate
+  # command-line argument for `kexec`. Otherwise nil.
+  def maybe_dtb()
+    log("Looking for a DTB file...")
+
+    mapping_path = generation_file("mobile-nixos/dtb-mapping.json", missing_allowed: true)
+    unless mapping_path && File.exist?(mapping_path)
+      log("  DTB: dtb-mapping.json not found... skipping DTB mapping...")
+      return nil
+    end
+
+    log("  DTB: Mappings from: `#{mapping_path}`...")
+    mapping = JSON.parse(File.read(mapping_path))
+
+    # Work only off the first compatible name.
+    # It is assumed that the loader uses the exact same scheme as the kernel
+    # build does. If this assumption stops holding true, it will be a new
+    # feature to implement.
+    board_compatible = File.read("/proc/device-tree/compatible").split("\0").first
+    log("  DTB: board_compatible: #{board_compatible}")
+
+    desired_dtb = mapping[board_compatible]
+    if desired_dtb
+      log("  DTB: wants: #{desired_dtb}")
+    else
+      log("  DTB: no DTB mapping found for #{board_compatible}...")
+      return nil
+    end
+
+    # The desired_dtb path is an absolute path in the mounted system.
+    file = File.join(SYSTEM_MOUNT_POINT, desired_dtb)
+
+    if File.exist?(file)
+      log("  DTB: file `#{file}` found")
+      "--dtb=#{forward_fdt_bootloader_info(file)}"
+    else
+      log("  DTB: file `#{file}` not found... skipping DTB mapping")
+      nil
+    end
+  end
+
+  # Given a path to a DTB file, merges required properties that the bootloader
+  # has setup. It will additionally merge optional properties.
+  def forward_fdt_bootloader_info(path)
+    args = [
+      "--print-header",
+      "--copy-dtb", path,
+      Configuration["stage-0"]["forward"]["nodes"].map {|path| [ "--forward-node", path] },
+      Configuration["stage-0"]["forward"]["props"].map {|pair| [ "--forward-prop", *pair] },
+    ].flatten
+    log(" $ fdt-forward #{args.shelljoin}")
+    dts = `fdt-forward #{args.shelljoin}`
+
+    # Declare we booted using stage-0's kexec
+    # And additional useful debugging data...
+    dts << [
+      "\n",
+      "// Declare we booted using kexec",
+      %Q{/ { mobile-nixos,stage-0; };},
+      %Q{/ { mobile-nixos,stage-0,timestamp = #{Time.now.to_s.to_json}; };},
+      %Q{/ { mobile-nixos,stage-0,uname = #{`uname -a`.to_json}; };},
+      %Q{/ { mobile-nixos,stage-0,uptime = #{`uptime`.to_json}; };},
```

Comment:
`to_json`?

Because it will produce and escape strings well enough for use inside DTS files. The required escapes are basically compatible.